File: autocomplete_system_1.py
from collections import defaultdict
from typing import List


class AutocompleteSystem:

    def __init__(self, sentences: List[str], times: List[int]):
        """
            Time Complexity - O(mn)
            'm' is the average length of the string in the sentences list
            'n' is the number of sentences
            Space Complexity - O(n)
        """
        self.dic = defaultdict(int)
        for i in range(len(sentences)):
            self.dic[sentences[i]] += times[i]
        self.search = ''

    def input(self, c: str) -> List[str]:
        """
            Time Complexity - O(n+mlogm)
            'm' the number of strings in the result list
            'n' is the number of sentences
            Space Complexity - O(m)
        """
        result = []

        if c == '#':
            self.dic[self.search] += 1
            self.search = ''
            return result

        self.search += c
        for k, v in self.dic.items():
            if k.startswith(self.search):
                result.append((k, -v))
        result.sort(key=lambda x: (x[1], x[0]))
        return [result.pop(0)[0] for _ in range(min(len(result), 3))]


# Your AutocompleteSystem object will be instantiated and called as such:

if __name__ == '__main__':
    """
    ["AutocompleteSystem","input","input","input","input","input","input","input","input","input","input","input","input"]
[[,],["i"],[" "],["a"],["#"],["i"],[" "],["a"],["#"],["i"],[" "],["a"],["#"]]
    """
    obj = AutocompleteSystem(["i love you", "island", "iroman", "i love leetcode"], [5, 3, 2, 2])
    print(obj.input("i"))
    print(obj.input(" "))
    print(obj.input("a"))
    print(obj.input("#"))
    print(obj.input("i"))
    print(obj.input(" "))
    print(obj.input("a"))
    print(obj.input("#"))
    print(obj.input("i"))
    print(obj.input(" "))
    print(obj.input("a"))
    print(obj.input("#"))
    """
    ["i love you","island","i love leetcode"],
    ["i love you","i love leetcode"],
    [],
    []
    
    ["i love you","island","i love leetcode"],
    ["i love you","i love leetcode","i a"],
    ["i a"],
    []
    
    ["i love you","island","i a"],
    ["i love you","i a","i love leetcode"],
    ["i a"],
    []]
    """

File: AutoCompleteSystem1.java
// Time Complexity : O(n) -> list of the words of the prefix that the user types
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes

//with Trie and PriorityQueue with all the available strings of prefixes in trie

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.PriorityQueue;

//With Trie and PriorityQueue
public class AutoCompleteSystem1 {
        HashMap<String, Integer> map;
        String search;
        TrieNode root;
        class TrieNode{
            TrieNode[] children;
            List<String> startsWith;

            public TrieNode(){
                children = new TrieNode[256];
                startsWith = new ArrayList<>();
            }
        }
        public AutoCompleteSystem1(String[] sentences, int[] times) { // O(n*l)  insertion into trie by programmer
            map = new HashMap<>();
            root = new TrieNode();
            search = "";
            for(int i=0; i<sentences.length; i++)
            {
                if(!map.containsKey(sentences[i])) insert(sentences[i]);
                map.put(sentences[i],map.getOrDefault(sentences[i],0)+times[i]);
            }
        }

        public void insert(String word){
            TrieNode curr = root;
            for(int i=0; i<word.length(); i++){
                char c = word.charAt(i);
                if(curr.children[c-' ']==null){
                    curr.children[c-' '] = new TrieNode();
                }
                curr = curr.children[c- ' '];
                curr.startsWith.add(word);
            }
        }
        public List<String> searchPrefix(String prefix){  //O(l)  l is the length of the prefix
            List<String> result = new ArrayList<>();
            TrieNode curr = root;
            for(int i=0; i<prefix.length(); i++)
            {
                char c = prefix.charAt(i);
                if(curr.children[c-' ']== null){
                    return result;
                }
                curr = curr.children[c-' '];
            }
            return curr.startsWith;
        }
        public List<String> input(char c) {
            List<String> result = new ArrayList<>();
            if(c == '#')
            {
                if(!map.containsKey(search)){
                    insert(search);
                }
                map.put(search, map.getOrDefault(search,0)+1);
                search = "";
                return result;
            }
            search += c;
            PriorityQueue<String> pq = new PriorityQueue<>((String a, String b)->{
                if(map.get(a) == map.get(b)){
                    return b.compareTo(a);
                }
                return map.get(a) - map.get(b);
            });
            List<String> wordsMatch = searchPrefix(search);
            for(String sentence: wordsMatch)   //O(n)
            {
                pq.add(sentence);
                if(pq.size() > 3) pq.poll();
            }
            while(!pq.isEmpty()){
                result.add(0,pq.poll());
            }
            return result;
        }
    }

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem obj = new AutocompleteSystem(sentences, times);
 * List<String> param_1 = obj.input(c);
 */
